### PR TITLE
fix: auto-detect and recover corrupted or empty worktree directories

### LINKED ISSUE #190: fix: auto-detect and recover corrupted or empty worktree directories
## Problem

Oompa's git worktrees occasionally end up in a corrupted or empty state, causing operations to fail with:

\`\`\`
fatal: not a git repository (or any parent up to mount point /)
\`\`\`

This has been observed across multiple repos:
- **nmstate/kubernetes-nmstate triage worktree**: Empty directory (no \`.git\` file), caused triage investigation to fail silently
- **ovn-kubernetes/ovn-kubernetes**: Corrupted git objects (\`unable to read sha1 file\`), caused rebase failures every 2 minutes
- **openshift/hypershift**: Corrupted git objects, caused a review retry loop burning ~\$2,000+ over 5 days

When a worktree is broken, oompa retries the failing operation every poll cycle without ever recovering, wasting agent invocations and API calls.

## Root Cause

Worktrees can become corrupted when:
1. Oompa is killed mid-operation (e.g., during version update restart while a git operation is in progress)
2. The base repo diverges significantly from the worktree branch (upstream file deletions cause \`unable to read sha1\` errors)
3. Disk space issues on tmpfs
4. A worktree directory is created but the git operation to initialize it fails partway through, leaving an empty directory

### 1. Validate worktrees before use

Before any git operation in a worktree, verify it's healthy:

\`\`\`go
func (g *GitWorktreeManager) isWorktreeHealthy(ctx context.Context, worktreePath string) bool {
    // Check .git file exists (worktrees have a .git FILE, not directory)
    gitPath := filepath.Join(worktreePath, ".git")
    info, err := os.Stat(gitPath)
    if err != nil || info.IsDir() {
        return false // missing or is a directory (should be a file for worktrees)
    }

    // Check git status works
    _, _, err = g.runner.Run(ctx, worktreePath, "git", "status")
    return err == nil
}
\`\`\`

### 2. Auto-recover unhealthy worktrees

When a worktree is detected as unhealthy, remove and recreate it:

\`\`\`go
func (g *GitWorktreeManager) ensureWorktreeHealthy(ctx context.Context, worktreePath, branchName string) error {
    if g.isWorktreeHealthy(ctx, worktreePath) {
        return nil
    }

    g.logger.Warn("detected unhealthy worktree, recreating",
        "path", worktreePath,
        "branch", branchName)

    // Remove the broken worktree
    os.RemoveAll(worktreePath)
    g.runner.Run(ctx, g.cloneDir, "git", "worktree", "prune")

    // Recreate
    return g.CreateWorktree(ctx, branchName)
}
\`\`\`

### 3. Validate base repo clone health

In \`EnsureRepoCloned\`, after confirming \`.git\` exists, also run a quick health check:

\`\`\`go
// After checking .git exists (line 75):
_, _, err := g.runner.Run(ctx, g.cloneDir, "git", "rev-parse", "HEAD")
if err != nil {
    // Base repo is corrupted -- nuke and re-clone
    g.logger.Warn("base repo corrupted, re-cloning", "path", g.cloneDir)
    os.RemoveAll(g.cloneDir)
    // Fall through to the clone path below
}
\`\`\`

### 4. Add health check to ensureWorktreeReady

The existing \`ensureWorktreeReady\` method (called before agent invocations) should call \`ensureWorktreeHealthy\` to auto-recover:

\`\`\`go
func (a *Agent) ensureWorktreeReady(ctx context.Context, work *IssueWork) error {
    if err := a.worktrees.EnsureRepoCloned(ctx); err != nil {
        return err
    }
    return a.worktrees.ensureWorktreeHealthy(ctx, work.WorktreePath, work.BranchName)
}
\`\`\`

### 5. Log corruption events as errors for Slack reporting

When a corrupted worktree is detected and recovered, emit an error-category event so it shows up in Slack reports:

\`\`\`go
a.emit(Event{
    Type:     EventError,
    Category: CategoryError,
    Worker:   a.workerName(),
    Action:   "Recovered corrupted worktree",
    Detail:   fmt.Sprintf("PR #%d worktree was unhealthy, recreated", work.PRNumber),
})
\`\`\`

## What this prevents

| Scenario | Before | After |
|----------|--------|-------|
| Empty worktree directory | Silent failure every cycle | Auto-recreate, one-time recovery |
| Corrupted git objects | Retry loop (rebase/review fails every 2min) | Nuke and re-clone, one-time recovery |
| Corrupted base repo | All worktrees fail | Re-clone base repo |
| Kill during git operation | Orphaned partial state | Detected and cleaned on next cycle |

## Tests

1. Empty worktree directory (no \`.git\` file) → detected as unhealthy → removed and recreated
2. Worktree with \`.git\` file but \`git status\` fails → detected as unhealthy → removed and recreated
3. Healthy worktree → passes validation → no action taken
4. Base repo \`.git\` exists but \`git rev-parse HEAD\` fails → re-cloned
5. Recovery emits error event for Slack reporting
6. After recovery, the operation that triggered it succeeds on retry

## Files to change

| File | Change |
|------|--------|
| \`pkg/agent/worktree.go\` | Add \`isWorktreeHealthy\`, \`ensureWorktreeHealthy\`, base repo health check in \`EnsureRepoCloned\` |
| \`pkg/agent/loop.go\` | Update \`ensureWorktreeReady\` to call health check |
| \`pkg/agent/triage.go\` | Use \`ensureWorktreeHealthy\` before triage agent invocation |
| \`pkg/agent/worktree_test.go\` | Test corruption detection and recovery |

### PR DESCRIPTION
**Pull request by qinqon**
Fixes #190

## Summary

Auto-detect and recover corrupted or empty worktree directories that cause infinite retry loops burning agent invocations and API costs. When a worktree is unhealthy (missing `.git` file, corrupted git objects, or failed `git status`), it is automatically removed and recreated on the next poll cycle instead of failing repeatedly.

## Changes

- Add `isWorktreeHealthy` method to `GitWorktreeManager` that validates a worktree by checking for a `.git` file (not directory) and running `git status`
- Add `EnsureWorktreeHealthy` method to `GitWorktreeManager` (and `WorktreeManager` interface) that auto-recovers unhealthy worktrees by removing and recreating them
- Add base repo corruption detection in `EnsureRepoCloned`: runs `git rev-parse HEAD` after confirming `.git` exists, and re-clones if the repo is corrupted
- Update `ensureWorktreeReady` in `git_ops.go` to call `EnsureWorktreeHealthy` before `CreateWorktree`, so corrupted worktrees are recovered before any agent invocation
- Add comprehensive tests for all corruption scenarios (empty directory, corrupted git status, healthy worktree, corrupted base repo)

## Testing

- [x] `go test ./...` passes
- [x] `go vet ./...` passes
- [x] Manual testing (if applicable)

## Related Issues

Fixes #190

## Summary by CodeRabbit

* **New Features**
  * Added automatic health validation for git worktrees
  * Automatic detection and recovery from corrupted worktree states
  * Automatic detection and recovery from corrupted repository states

* **Tests**
  * Added comprehensive test coverage for worktree health checks and recovery scenarios

[Review Change Stack](https://app.coderabbit.ai/change-stack/qinqon/oompa/pull/191?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: pkg/agent/git_ops.go [hunk 1]
```diff
@@ -48,10 +48,19 @@ func stripSignedOffBy(s string) string {
...
 		return err
 	}
+
+	// Check if existing worktree is healthy; recover if not
+	if work.WorktreePath != "" {
+		if err := a.worktrees.EnsureWorktreeHealthy(ctx, work.WorktreePath, work.BranchName); err != nil {
+			return err
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/git_ops.go]: > This block is redundant because `CreateWorktree` is already responsible for ensuring a worktree is ready for use. By integrating the health check directly into the `CreateWorktree` reuse logic, you can simplify this function and avoid redundant lock acquisitions and duplicate cleanup steps.

Addressed: Removed the `EnsureWorktreeHealthy` call from `ensureWorktreeReady`. The health check is now integrated directly into `CreateWorktree`'s reuse path, so `ensureWorktreeReady` simply calls `CreateWorktree` which handles health validation internally.

File: pkg/agent/worktree.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ type WorktreeManager interface {
 	CreateWorktree(ctx context.Context, branchName string) (worktreePath string, err error)
 	RemoveWorktree(ctx context.Context, worktreePath string) error
 	SyncWorktree(ctx context.Context, worktreePath string) error
+	EnsureWorktreeHealthy(ctx context.Context, worktreePath, branchName string) error
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/worktree.go]: > `EnsureWorktreeHealthy` is unnecessary as an exported interface method. Its functionality is better suited as an internal check within `CreateWorktree` to determine if an existing worktree can be safely reused.

Addressed: Removed `EnsureWorktreeHealthy` from the `WorktreeManager` interface. The health check (`isWorktreeHealthy`) is now an internal method called directly within `CreateWorktree`'s reuse path — if a worktree has a `.git` file but `git status` fails, it falls through to the cleanup+recreate path.

File: pkg/agent/worktree.go [hunk 3]
```diff
@@ -204,3 +213,34 @@ func (g *GitWorktreeManager) RemoveWorktree(ctx context.Context, worktreePath st
...
+	}
+
+	// Remove the broken worktree
+	os.RemoveAll(worktreePath) //nolint:errcheck // best-effort
+	g.runner.Run(ctx, g.cloneDir, "git", "worktree", "prune") //nolint:errcheck // best-effort
+
+	// Recreate via CreateWorktree (which handles lock, cleanup, and branch creation)
+	_, err := g.CreateWorktree(ctx, branchName)
+	return err
+}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/worktree.go]: > This method can be removed if the health check is integrated into `CreateWorktree`. Note that it also performs manual cleanup (`os.RemoveAll` and `git worktree prune`) that is already handled by `CreateWorktree` (lines 165-167), leading to duplicate operations.

Addressed: Removed `EnsureWorktreeHealthy` entirely. The health check is now integrated into `CreateWorktree`'s reuse path via `isWorktreeHealthy`, and cleanup is handled by `CreateWorktree`'s existing cleanup logic — no more duplicate `os.RemoveAll`/`prune` operations.

File: pkg/agent/loop_test.go [hunk 4]
```diff
@@ -273,6 +273,10 @@ func (m *mockWorktreeManager) RemoveWorktree(_ context.Context, worktreePath str
 	return nil
 }
 
+func (m *mockWorktreeManager) EnsureWorktreeHealthy(_ context.Context, _, _ string) error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/loop_test.go]: > This mock implementation can be removed once the `EnsureWorktreeHealthy` method is removed from the `WorktreeManager` interface.

Addressed: Removed this mock method along with the interface method it implemented.

File: pkg/agent/worktree.go [hunk 5]
```diff
@@ -204,3 +213,34 @@ func (g *GitWorktreeManager) RemoveWorktree(ctx context.Context, worktreePath st
...
+// Returns nil if the worktree is healthy or was successfully recovered.
+func (g *GitWorktreeManager) EnsureWorktreeHealthy(ctx context.Context, worktreePath, branchName string) error {
+	if g.isWorktreeHealthy(ctx, worktreePath) {
+		return nil
+	}
+
+	// Remove the broken worktree
+	os.RemoveAll(worktreePath) //nolint:errcheck // best-effort
+	g.runner.Run(ctx, g.cloneDir, "git", "worktree", "prune") //nolint:errcheck // best-effort
+
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/worktree.go]: > `EnsureWorktreeHealthy` performs filesystem deletion and runs `git worktree prune` without taking `g.mu`

Addressed: `EnsureWorktreeHealthy` has been removed entirely. Per other review feedback, the health check was integrated directly into `CreateWorktree`'s reuse path, which already holds `g.mu`. The concern about unserialized mutations no longer applies.

File: pkg/agent/worktree_test.go [hunk 6]
```diff
@@ -221,3 +227,231 @@ func TestEnsureRepoCloned_Fresh(t *testing.T) {
...
+	}
+	if !foundWorktreeAdd {
+		t.Error("expected 'git worktree add' to recreate unhealthy worktree")
+	}
+
+	// Verify the empty directory was removed
+	if _, err := os.Stat(worktreePath); err == nil {
+		// It's OK if it exists — CreateWorktree (mock) may have recreated it.
+		// The key assertion is that prune + add were called.
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/worktree_test.go]: > Remove the empty `if` block causing SA9003 lint failure.

Already addressed: The `TestEnsureWorktreeHealthy_EmptyDirectoryRecreates` test (which contained this empty if block) was replaced with `TestCreateWorktree_EmptyDirectoryRecreates` in the previous commit, which does not have the empty branch. The SA9003 issue no longer exists in the current code.

File: pkg/agent/worktree.go [hunk 7]
```diff
@@ -204,3 +213,34 @@ func (g *GitWorktreeManager) RemoveWorktree(ctx context.Context, worktreePath st
...
+	// Recreate via CreateWorktree (which handles lock, cleanup, and branch creation)
+	_, err := g.CreateWorktree(ctx, branchName)
+	return err
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/worktree.go]: > `EnsureWorktreeHealthy` mutates worktree/git state (`os.RemoveAll`, `git worktree prune`) without acquiring `g.mu`

Already addressed: `EnsureWorktreeHealthy` was removed entirely in the previous commit. The health check (`isWorktreeHealthy`) is now integrated directly into `CreateWorktree`'s reuse path, which already holds `g.mu`. All git-mutating operations are properly serialized under the existing mutex.
